lib/src/rules/avoid_leading_underscore_for_non_private_identifiers.dart

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2015, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/utils.dart';
+
+const _desc = r'Avoid leading underscore for identifiers that aren’t private.';
+
+const _details = r'''
+
+**DON’T** use a leading underscore for identifiers that aren’t private.
+
+Dart uses a leading underscore in an identifier to mark members and top-level declarations as private. This trains users to associate a leading underscore with one of those kinds of declarations. They see “_” and think “private”.
+
+There is no concept of “private” for local variables, parameters, or library prefixes. When one of those has a name that starts with an underscore, it sends a confusing signal to the reader. To avoid that, don’t use leading underscores in those names.
+
+**Exception**: An unused parameter can be named _, __, ___, etc. This happens in things like callbacks where you are passed a value but you don’t need to use it. Giving it a name that consists solely of underscores is the idiomatic way to indicate the value isn’t used.
+
+**BAD**
+```
+print(string _name) {
+  int _size = _name.length;
+}
+```
+
+**GOOD:**
+```
+print(string name) {
+  int size = name.length;
+}
+```
+
```

Comment:
maybe consider adding an "OK" example for the unused parameter case?

Reply:
Added OK case for unused callback parameter.